### PR TITLE
PWGHF: ML selection in LcToK0sP candidate selector

### PR DESCRIPTION
**Pull request by DanielSamitz**
Implementation of possibility to use ONNX ML models in the candidate selector for Lc->K0sP

In `inputParamsAll` in https://github.com/DanielSamitz/O2Physics/blob/3a0cae428cb9ad827d1aa0e5e9fea0b2698d9d1f/PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx#L208C11-L232 I'm filling all possible variables that could be used as inputs for a ML model into an array, and in https://github.com/DanielSamitz/O2Physics/blob/3a0cae428cb9ad827d1aa0e5e9fea0b2698d9d1f/PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx#L237-L243 only the ones selected via the configurable `MLInputs` are filled into the vector that is eventually passed to the model. Might be a bit of an overkill, but in this way one is completely flexible in the input features of the models (no hard coded inputs). One just has to make sure that the list of 'all' possible variables in `inputParamsAll` is well chosen. Not too many (because these variables have to be calculated and filled into the array for each candidate that passes the pre-selection cuts), but still covering any reasonable choice of input variables that one might choose.

Left for later work:

- retrieving models from ccdb
- possibility for different models in the different pT bins (for the moment only different thresholds possible)

### CODE REVIEW

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:1]: Change `BDT` -> `Bdt` in variables throughout the code.

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:25]: Move before PWGHF

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:68]: Move down, before `init`

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:238]: Use `TESTBIT`

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:242]: Where does the correspondence between the position of `param` and `i` come from?
Why don't you loop over `i` and push `inputParamsAll[i]`?

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:246]: Where are the values of `dataTypeML` documented?

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:271]: The names of the columns should be fixed.

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:280]: Is this needed?

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:306]: Why don't you use one `if`?

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:339]: Same as above?

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:1]: OK (changed variable names, left it in text)

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: OK, done

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: OK, done

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: OK, done

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: No reason. Fixed it as you suggested

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:246]: Here https://onnx.ai/onnx/api/mapping.html#l-onnx-types-mapping or also here https://github.com/microsoft/onnxruntime/blob/961fa7274ad1631bb5364ef6f836d99558b9d3b0/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api.h#L46-L68

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: No, it was just meant to be as a placeholder for later, because for the moment there is no distinction between 'prompt' and 'nonprompt'. It only returns 'false' for background and 'true' for signal (prompt or nonprompt). I took it out

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: I was not sure if `&&` guarantees that the right part never gets called if the left evaluates to 'false'. Fixed now as you suggested

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: The names of the columns are defined in https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/Core/SelectorCuts.h, and are also used in https://github.com/AliceO2Group/O2Physics/blob/master/EventFiltering/PWGHF/HFFilterHelpers.h and https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorLcMl.cxx
Should I change them everywhere (to BdtBkg, BdtPrompt, BdtNonPrompt?) or do this in a separate PR?

**Comment by DanielSamitz** [comment]
Only now I saw https://github.com/AliceO2Group/O2Physics/pull/2841.
Maybe I should first wait for that to proceed and be merged, and then use what I can from the new class so that we have everything in HF on the same basis

**Comment by fcatalan92** [comment]
Hi @DanielSamitz, yes, please. IMHO it is worth waiting a bit and using the HF dedicated class that is under development, which actually will solve also your current todos (ccdb access, different model for different pT bins).

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx:246]: OK, thanks. Would be quite nice, if once could use the `DataType` names directly. Would it be complicated to make the namespace visible in O2?

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (vkucera) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: Yes, it does. Thanks

**Comment by vkucera** [comment]
Hi @DanielSamitz , I have approved https://github.com/AliceO2Group/O2Physics/pull/2841. Once it gets merged, please adapt your PR accordingly and mark your PR as ready for review when you're done.

**Comment by DanielSamitz** [comment]
OK @vkucera , thanks for the info.

**Comment by DanielSamitz** [comment]
New version of the ONNX model inference using PWGHF/Core/HfMlResponse, and passing a configurable list of strings with the names of the required input parameters of the model.

**Comment by DanielSamitz** [comment]
@vkucera sorry for the force-push (unfortunately I did a rebase  to merge with the master branch), but let's say it is anyways 95% new with respect to the last reviewed commit.

File: PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateSelectorLcToK0sP.cxx]: I used an `enum` and a `std::map` for the names of the input features. In this way one is also safe from messing up the order of the names compared to the order in the enum.

**Review by vkucera — APPROVED** [review]
Thanks @DanielSamitz and sorry for the delay.

**Pull request merged**
